rancher/rancher
# Fix migrator

## Issue: 

 https://github.com/rancher/rancher/issues/38699

## Problems

Previously, there were a few problems with how objects were being passed and used. Conditions were being set on clusters, but the change was not committed with a client Update call.

A Do function would pass in one pointer but then return a different pointer value pointing to a more updated cluster object. The code would then proceed to use the outdated object. This would cause the service account token secret migration update to fail here: https://github.com/rancher/rancher/blob/release/v2.6/pkg/controllers/management/secretmigrator/clusters.go#L661. This was not a problem for some types of clusters like AKSV2 clusters because there is redundant code for migrating the service account token in their respective sync handlers and the migration code in clusters.go would short circuit before the problematic code was hit (see: https://github.com/rancher/rancher/blob/release/v2.6/pkg/controllers/management/aks/aks_cluster_handler.go#L214 and https://github.com/rancher/rancher/blob/release/v2.6/pkg/controllers/management/secretmigrator/clusters.go#L650).
    
Modifications were made to the object referenced by the pointer passed into the sync function. This is bad since this object will be passed to subsequent handlers in the case that a nil cluster object is returned. Now, the originally referenced object is not modified.

## Solutions

Now, condition changes are followed by an update.

Now, the returned object and object that was passed in are set to have the same pointer value which is up to date.

Now, the originally referenced object is not modified.

## Testing

To reproduce this issue:
1. Run rancher v2.6.3
2. Navigate to `<server-url>/g/clusters/add/launch/azureaks`
3. Create AKSv1 cluster using the UI.
4. Wait for cluster to become active.
5. Upgrade to v2.6.7.
6. Try to access workloads in system project of AKSv1 cluster. Agent will be outdated.
7. View rancher logs. Should see logs stating it is unavailable.

## Engineering Testing
### Manual Testing

Followed the above steps and ensured expected behavior has been restored.

### Automated Testing

Added unit tests to ensure cluster is being updated by client, conditions are being set, secrets are migrated.

## QA Testing Considerations

QA should tests that the upgrade works as expected for AKSV1 and AKSV2 clusters.
 
### Regressions Considerations
